{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6b56da8a-477c-45d2-b96c-26934a9f5f57",
   "metadata": {},
   "source": [
    "\n",
    "## **Tool: Unicycler**\n",
    "\n",
    "**Authors: Ryan R. Wick, Louise M. Holt, et al.**\n",
    "\n",
    "\n",
    "**Year: 2017**\n",
    "\n",
    "\n",
    "\n",
    "### **Use of the tool:** ###\n",
    "\n",
    "Unicycler is a bioinformatics tool used for hybrid assembly of bacterial genomes from short-read sequencing data. It combines both de novo assembly and read mapping approaches to produce highly accurate and complete genome assemblies. Unicycler is specifically designed for handling bacterial genomes, including those with repetitive regions, plasmids, and complex genomic structures.\n",
    "\n",
    "#### **Distinguishing features:** ####  \n",
    "\n",
    "Unicycler differs from other assembly tools by integrating the benefits of both de novo assembly and read mapping. It first constructs an initial assembly graph using de Bruijn graph-based assembly, followed by read mapping to resolve repetitive regions and correct errors. This hybrid approach results in highly accurate and complete bacterial genome assemblies.\n",
    "\n",
    "#### **How the tool works:** #### \n",
    "\n",
    "Unicycler works in several steps. It starts by performing a preliminary de novo assembly using the SPAdes assembler. This step generates contigs, including circular contigs representing plasmids or other circular genomic elements. The tool then uses a combination of read mapping and iterative correction to improve the assembly. It maps the reads back to the contigs, identifies repetitive regions, and uses short reads to correct errors. The process is repeated until the assembly is refined and circularized.\n",
    "\n",
    "#### **Installation:** #### \n",
    "``` bash\n",
    "conda install -c bioconda unicycler\n",
    "\n",
    "```\n",
    "\n",
    "#### **Input files:** #### \n",
    "\n",
    "To run Unicycler, you need short-read sequencing data in FASTQ format. The tool accepts both paired-end and single-end reads. Additionally, you may provide long-read data in FASTQ or FASTA format for improved assembly results.\n",
    "\n",
    "#### **Standard use:** #### \n",
    "\n",
    "``` bash \n",
    "unicycler -1 <forward_reads.fastq> -2 <reverse_reads.fastq> -o <output_directory> --min_fasta_length 300 -t 85 --keep 1\n",
    "\n",
    "```\n",
    "\n",
    "##### **Options used in the command:** #####\n",
    "\n",
    "\n",
    "- ` -1 <forward_reads.fastq>: Specifies the input file containing forward reads in FASTQ format.`\n",
    "- ` -2 <reverse_reads.fastq>: Specifies the input file containing reverse reads in FASTQ format.`\n",
    "- ` -o <output_directory>: Specifies the output directory where the results will be saved.`\n",
    "- ` --min_fasta_length: Specify the minimum length of contigs to be included in the assembly.`\n",
    "- ` --keep 1: Specify which intermediate files to keep after the assembly process.`\n",
    "\n",
    "\n",
    "##### **Additional options:** ##### \n",
    "\n",
    "Unicycler provides additional options to fine-tune the assembly process, such as specifying long-read data, changing the assembly mode, adjusting the number of threads, etc. You can refer to the tool's documentation for a comprehensive list of options and their usage.\n",
    "\n",
    "\n",
    "##### **Results:** ##### \n",
    "\n",
    "- ` The output result files of Unicycler include:`\n",
    "- ` Assembly.fasta: The final assembly in FASTA format.`\n",
    "- ` Assembly.gfa: The assembly graph in GFA format.`\n",
    "- ` Assembly.log: A log file containing information about the assembly process.`\n",
    "- ` Subdirectories containing intermediate files and results.`\n",
    "\n",
    "##### **Structure of important result files:** #####\n",
    "\n",
    "Assembly.fasta: This file contains the final assembled genome, including the main chromosome and any plasmids or circular genomic elements. It can be used for downstream analyses, such as gene prediction, annotation, comparative genomics, etc.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1abbf204-9e2c-4217-a923-156408614e41",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate bioinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b665add4-2363-4b77-9439-d3ed21645d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "rawpath=\"/home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Data\"\n",
    "\n",
    "# out_dir=\"/media/chrf/GenomeBack/SPN/All_Batchs_SPN\"\n",
    "\n",
    "\n",
    "for files in $rawpath\\/*\n",
    "do\n",
    "    name=`basename $files| cut -f 1,2 -d '_'`\n",
    "\n",
    "    echo $name\n",
    "    mkdir -p -v $rawpath\\/$name\\/1_RawData\n",
    "    mv $rawpath\\/$name*\\_R1_001.fastq.gz $rawpath\\/$name*\\_R2_001.fastq.gz $rawpath\\/$name\\/1_RawData/\n",
    "    unicycler -1 $rawpath\\/$name\\/1_RawData\\/$name*\\_R1*.fastq.gz -2 $rawpath\\/$name\\/1_RawData\\/$name*\\_R2*.fastq.gz --min_fasta_length 300 -t 85 --keep 1 -o $rawpath\\/$name\\/3_Unicycler/\n",
    "    mv $rawpath\\/$name\\/3_Unicycler\\/assembly.fasta $rawpath\\/$name\\/3_Unicycler\\/$name\\_contigs.fasta\n",
    "    break\n",
    "\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
