{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f45166d7-ac9b-4471-abe5-90c3a31538f7",
   "metadata": {
    "tags": []
   },
   "source": [
    "![Cat](logo.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b56da8a-477c-45d2-b96c-26934a9f5f57",
   "metadata": {
    "tags": []
   },
   "source": [
    "\n",
    "## **Tool: Unicycler**\n",
    "\n",
    "**Authors: Ryan R. Wick, Louise M. Holt, et al.**\n",
    "\n",
    "\n",
    "**Year: 2017**\n",
    "\n",
    "\n",
    "\n",
    "### **Use of the tool:** ###\n",
    "\n",
    "\n",
    ">- **Unicycler is a bioinformatics tool used for**\n",
    "\n",
    ">>- **de-novo assembly of bacterial genomes from short-read sequencing data.** \n",
    ">>- **It combines both de novo assembly and read mapping approaches to produce highly accurate and complete genome assemblies.**\n",
    "\n",
    "\n",
    "\n",
    "### **How the tool works:** ###\n",
    "\n",
    ">- **Unicycler works in several steps.**\n",
    ">- **It starts by performing a preliminary de novo assembly using the SPAdes assembler.**\n",
    ">>- **This step generates contigs, including circular contigs representing plasmids or other circular genomic elements.**\n",
    "\n",
    "\n",
    "\n",
    "### **Installation:** ###\n",
    "\n",
    "``` bash\n",
    "conda install -c bioconda unicycler\n",
    "\n",
    "```\n",
    "\n",
    "### **Input files:** ###\n",
    "\n",
    ">- **To run Unicycler,**\n",
    ">>- **you need short-read sequencing data in FASTQ format.**\n",
    ">>- **The tool accepts both paired-end and single-end reads.**\n",
    "\n",
    "\n",
    "### **Standard use:** ###\n",
    "\n",
    "``` bash \n",
    "unicycler -1 <forward_reads.fastq> -2 <reverse_reads.fastq> -o <output_directory> --min_fasta_length 300 -t 16 --keep 1\n",
    "\n",
    "```\n",
    "\n",
    "##### **Options used in the command:** #####\n",
    "\n",
    "\n",
    "- ` -1 <forward_reads.fastq>: Specifies the input file containing forward reads in FASTQ format.`\n",
    "- ` -2 <reverse_reads.fastq>: Specifies the input file containing reverse reads in FASTQ format.`\n",
    "- ` -o <output_directory>: Specifies the output directory where the results will be saved.`\n",
    "- ` --min_fasta_length: Specify the minimum length of contigs to be included in the assembly.`\n",
    "- ` --keep 1: Specify which intermediate files to keep after the assembly process.`\n",
    "\n",
    "\n",
    "\n",
    "#### **Results:** ####\n",
    "\n",
    "- ` The output result files of Unicycler include:`\n",
    "- ` Assembly.fasta: The final assembly in FASTA format.`\n",
    "- ` Assembly.gfa: The assembly graph in GFA format.`\n",
    "- ` Assembly.log: A log file containing information about the assembly process.`\n",
    "- ` Subdirectories containing intermediate files and results.`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1abbf204-9e2c-4217-a923-156408614e41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(bioinfo) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda activate bioinfo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2eaeace-97cc-43ce-8852-b855d7aedbbd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assembly_unicycler.ipynb  ERR2663496_2.fastq.gz    slide_picture\n",
      "assembly_unicycler.sh     Genome_Assembly_v2.pptx  Unicycler_Result\n",
      "ERR2663496_1.fastq.gz     logo.png                 unicycler_v1.pptx\n",
      "(bioinfo) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94407631-373f-4b6a-ae0a-0dfa6a3c7681",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "tput: No value for $TERM and no -T specified\n",
      "\n",
      "Starting Unicycler (2023-06-12 15:56:34)\n",
      "    Welcome to Unicycler, an assembly pipeline for bacterial genomes. Since you\n",
      "provided only short reads, Unicycler will essentially function as a SPAdes-\n",
      "optimiser. It will try many k-mer sizes, choose the best based on contig length\n",
      "and graph connectivity, and scaffold the graph using SPAdes repeat resolution.\n",
      "    For more information, please see https://github.com/rrwick/Unicycler\n",
      "\n",
      "Command: /usr/bin/unicycler -1 ERR2663496_1.fastq.gz -2 ERR2663496_2.fastq.gz -o Unicycler_Result --min_fasta_length 300 -t 16 --keep 1\n",
      "\n",
      "Unicycler version: v0.4.8\n",
      "Using 16 threads\n",
      "\n",
      "The output directory already exists and files may be reused or overwritten:\n",
      "  /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result\n",
      "\n",
      "Dependencies:\n",
      "  Program         Version           Status  \n",
      "  spades          3.13.1            good    \n",
      "  racon                             not used\n",
      "  makeblastdb     2.13.0+           good    \n",
      "  tblastn         2.13.0+           good    \n",
      "  bowtie2-build   2.5.1             good    \n",
      "  bowtie2         2.5.1             good    \n",
      "  samtools        1.17              good    \n",
      "  java            17.0.3-internal   good    \n",
      "  pilon           ?                 good    \n",
      "  bcftools                          not used\n",
      "\n",
      "SPAdes graph already exists. Will use this graph instead of running SPAdes:\n",
      "  /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/001_best_spades_graph.gfa\n",
      "\n",
      "\n",
      "Determining graph multiplicity (2023-06-12 15:56:35)\n",
      "    Multiplicity is the number of times a sequence occurs in the underlying\n",
      "sequence. Single-copy contigs (those with a multiplicity of one, occurring only\n",
      "once in the underlying sequence) are particularly useful.\n",
      "\n",
      "\n",
      "Cleaning graph (2023-06-12 15:56:35)\n",
      "    Unicycler now performs various cleaning procedures on the graph to remove\n",
      "overlaps and simplify the graph structure. The end result is a graph ready for\n",
      "bridging.\n",
      "\n",
      "Graph overlaps removed\n",
      "\n",
      "Removed zero-length segments:\n",
      "    76, 77, 78, 79, 83, 84, 87, 88, 89, 93, 96, 106\n",
      "\n",
      "Removed zero-length segments:\n",
      "    94, 97\n",
      "\n",
      "Merged small segments:\n",
      "    98, 99, 100, 101, 105, 107, 109, 110, 111, 112, 113, 114, 115, 116, 117,\n",
      "118, 119, 120, 121, 122, 123, 124\n",
      "\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/002_overlaps_removed.gfa\n",
      "\n",
      "    Unicycler now selects a set of anchor contigs from the single-copy contigs.\n",
      "These are the contigs which will be connected via bridges to form the final\n",
      "assembly.\n",
      "\n",
      "46 anchor segments (4,718,248 bp) out of 90 total segments (4,836,493 bp)\n",
      "\n",
      "\n",
      "Creating SPAdes contig bridges (2023-06-12 15:56:35)\n",
      "    SPAdes uses paired-end information to perform repeat resolution (RR) and\n",
      "produce contigs from the assembly graph. SPAdes saves the graph paths\n",
      "corresponding to these contigs in the contigs.paths file. When one of these\n",
      "paths contains two or more anchor contigs, Unicycler can create a bridge from\n",
      "the path.\n",
      "\n",
      "                                Bridge\n",
      "Start       Path       End     quality\n",
      "  -33        76        39         62.2\n",
      "  -20    -80 → -80     27         59.8\n",
      "   12   67 → 54 → 67   -21         4.8\n",
      "   31        76        17         63.1\n",
      "\n",
      "\n",
      "Creating loop unrolling bridges (2023-06-12 15:56:35)\n",
      "    When a SPAdes contig path connects an anchor contig with the middle contig\n",
      "of a simple loop, Unicycler concludes that the sequences are contiguous (i.e.\n",
      "the loop is not a separate piece of DNA). It then uses the read depth of the\n",
      "middle and repeat contigs to guess the number of times to traverse the loop and\n",
      "makes a bridge.\n",
      "\n",
      "                                  Loop count   Loop count    Loop    Bridge\n",
      "Start   Repeat   Middle     End    by repeat    by middle   count   quality\n",
      "   12       67       54     -21         0.74         1.03       1      44.7\n",
      "\n",
      "\n",
      "Applying bridges (2023-06-12 15:56:35)\n",
      "    Unicycler now applies to the graph in decreasing order of quality. This\n",
      "ensures that when multiple, contradictory bridges exist, the most supported\n",
      "option is used.\n",
      "\n",
      "Bridge type   Start → end   Path         Quality\n",
      "SPAdes           31 → 17    76            63.134\n",
      "SPAdes          -33 → 39    76            62.233\n",
      "SPAdes          -20 → 27    -80, -80      59.812\n",
      "loop             12 → -21   67, 54, 67    44.658\n",
      "\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/003_bridges_applied.gfa\n",
      "\n",
      "\n",
      "Bridged assembly graph (2023-06-12 15:56:35)\n",
      "    The assembly is now mostly finished and no more structural changes will be\n",
      "made. Ideally the assembly graph should now have one contig per replicon and no\n",
      "erroneous contigs (i.e a complete assembly). If there are more contigs, then\n",
      "the assembly is not complete.\n",
      "\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/004_final_clean.gfa\n",
      "\n",
      "Component   Segments   Links   Length      N50       Longest segment   Status    \n",
      "    total         82     126   4,836,983   204,168           488,731             \n",
      "        1         81     125   4,730,277   204,168           488,731   incomplete\n",
      "        2          1       1     106,706   106,706           106,706     complete\n",
      "\n",
      "\n",
      "Polishing assembly with Pilon (2023-06-12 15:56:35)\n",
      "    Unicycler now conducts multiple rounds of Pilon in an attempt to repair any\n",
      "remaining small-scale errors with the assembly.\n",
      "\n",
      "Aligning reads to find appropriate insert size range...\n",
      "Insert size 1st percentile:  237\n",
      "Insert size 99th percentile: 374\n",
      "\n",
      "Pilon polish round 1\n",
      "No Pilon changes\n",
      "\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/005_polished.gfa\n",
      "\n",
      "\n",
      "Rotating completed replicons (2023-06-12 16:10:11)\n",
      "    Any completed circular contigs (i.e. single contigs which have one link\n",
      "connecting end to start) can have their start position changed without altering\n",
      "the sequence. For consistency, Unicycler now searches for a starting gene (dnaA\n",
      "or repA) in each such contig, and if one is found, the contig is rotated to\n",
      "start with that gene on the forward strand.\n",
      "\n",
      "Segment   Length    Depth   Starting gene     Position   Strand    Identity   Coverage\n",
      "     18   106,706   1.16x   UniRef90_Q934T6     46,159   reverse     100.0%     100.0%\n",
      "\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/006_rotated.gfa\n",
      "\n",
      "\n",
      "Assembly complete (2023-06-12 16:10:25)\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/assembly.gfa\n",
      "Saving /home/asus/Desktop/CHRF_Project/Bioinformatics_Traning/Advance_Bioinformatics_Traning/Unicycler_Assembly/Unicycler_Result/assembly.fasta\n",
      "\n",
      "(bioinfo) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "unicycler \\\n",
    "    -1 ERR2663496_1.fastq.gz \\\n",
    "    -2 ERR2663496_2.fastq.gz \\\n",
    "    -o Unicycler_Result \\\n",
    "    --min_fasta_length 300 \\\n",
    "    -t 16 \\\n",
    "    --keep 1 "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
