{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94f1b986-d7cd-4158-a83c-4d47a31ac971",
   "metadata": {},
   "source": [
    "![Cat](fastqc.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20857f41-90c9-4e44-9c42-4280d800116a",
   "metadata": {},
   "source": [
    "## **FastQC** ##\n",
    "\n",
    "\n",
    ">- **A typical whole genome sequencing process involves**\n",
    ">>- **sample preparation,** \n",
    ">>- **library preparation** \n",
    ">>- **and sequencing.** \n",
    ">- **Errors occurring at each of these steps can negatively impact the quality of the sequence information.** \n",
    ">- **It is important to perform quality checks on the raw sequence reads before starting your analysis.**\n",
    "\n",
    "### **Some of the parameters for NGS data quality control.** ###\n",
    ">- **Check for primers and adaptors**\n",
    ">- **Trim low quality ends**\n",
    ">- **Remove low quality reads**\n",
    ">- **Remove short sequences**\n",
    ">- **Remove reads with a high proportion of ambiguous bases (“N”)**\n",
    ">- **Remove duplicates**\n",
    "\n",
    "\n",
    "\n",
    "## **Tool** ##\n",
    "\n",
    "**We will be using the FastQC tool to carry out quality control on reads.** \n",
    "\n",
    "\n",
    "## **Installation:** ## \n",
    "\n",
    "``` bash\n",
    "conda install -c bioconda fastqc\n",
    "```\n",
    "\n",
    "### **Input files:** ###\n",
    "\n",
    ">- **To run FastQC,**\n",
    ">>- **you need short-read sequencing data in FASTQ format.**\n",
    ">>- **The tool accepts both paired-end and single-end reads.**\n",
    "\n",
    "\n",
    "### **Standard use:** ###\n",
    "\n",
    "``` bash \n",
    "\n",
    "fastqc <forward_reads.fastq.gz>  <reverse_reads.fastq.gz> -o output_directory\n",
    "\n",
    "```\n",
    "\n",
    "##### **Options used in the command:** #####\n",
    "\n",
    "- `  <forward_reads.fastq.gz>: Specifies the input file containing forward reads in FASTQ format.`\n",
    "- `  <reverse_reads.fastq.gz>: Specifies the input file containing reverse reads in FASTQ format.`\n",
    "- ` -o <output_directory>: Specifies the output directory where the results will be saved.`\n",
    "\n",
    "\n",
    "#### **Results:** ####\n",
    "\n",
    " \n",
    "- ` <forward_reads_fastqc>.html`\n",
    "- ` <forward_reads_fastqc>.zip`\n",
    "- ` <reverse_reads_fastqc>.html`\n",
    "- ` <reverse_reads_fastqc>.zip`\n",
    "\n",
    "\n",
    "###### Let’s explore the **forward_reads_fastqc.html** ######\n",
    "\n",
    "\n",
    ">- **Basic Statistics**\n",
    ">- **Per base sequence quality** \n",
    ">- **Per tile sequence quality**\n",
    ">- **Per Sequence Quality Scores**\n",
    ">- **Per Base Sequence Content**\n",
    ">- **Per Sequence GC Content**\n",
    ">- **Per Base N Content**\n",
    ">- **Sequence Length Distribution**\n",
    ">- **Sequence Duplication Levels**\n",
    ">- **Overrepresented sequences**\n",
    ">- **Sequence Duplication Levels**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fdc4c2a-9e57-4a77-8886-a6a3722eba38",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqc /\n",
    "    <forward_reads.fastq.gz> /\n",
    "    <reverse_reads.fastq.gz> /\n",
    "    -o FastQC_Result\n",
    "    -f fastq /\n",
    "    -t 16 "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
